File: backtracking/combination_sum_ii.py
from typing import List


class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        self.result = []
        self.bt(nums, 0, [], target)
        return self.result

    def bt(self, nums, s_idx, path, rem):
        if rem == 0: return self.result.append(path)

        for i in range(s_idx, len(nums)):
            if i > s_idx and nums[i] == nums[i - 1]: continue
            if nums[i] > rem: break
            self.bt(nums, i + 1, path + [nums[i]], rem - nums[i])



class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        self.result = []
        nums.sort()
        self.bt(nums, 0, [], target)
        return self.result

    def bt(self, nums, s_idx, path, rem):
        if rem == 0: return self.result.append(path)

        used = set()
        for i in range(s_idx, len(nums)):
            if nums[i] > rem: continue
            if nums[i] in used: continue
            used.add(nums[i])
            self.bt(nums, i + 1, path + [nums[i]], rem - nums[i])



# 40. Combination Sum II
# https://leetcode.com/problems/combination-sum-ii/description/


# Approach 1: Backtracking
# 1. Combination Sum I allowed duplicates. To avoid those, in the recursive call, use i + 1,
#    in order to start from the next element
# 3. Now other duplicates are also possible. Imagine [1,2,5, 7, 1] and target as 8.
#    If we use DFS we will get [1,7] and then [7,1]. How do we avoid this?
# 4. Sort candidates: [1,1,2,5,7]. Now when you start with index 0,
#    your first element will be 1. It will allow you to pick the second element as 1 too.
#    You will be able to pick [1,7]. But during recursion, when you reach the next start index as 1,
#    your recursion tree will again start from 1. This will lead to a duplicate [1,7].
#    You want to avoid this.


# Another Implementation
# class Solution:
#     def combinationSum2(self, candidates, target):
#         """
#         :type candidates: List[int]
#         :type target: int
#         :rtype: List[List[int]]
#         """
#         all_solns = []
#         self.helper([], 0, sorted(candidates), target, all_solns)
#         return all_solns

#     def helper(self, so_far, k, nums, target, all_solns):
#         sum_so_far = sum(so_far)
#         if sum_so_far == target:
#             all_solns.append([x for x in so_far])
#         else:
#             for i in range(k, len(nums)):
#                 if i > k and nums[i] == nums[i-1]: continue
#                 if (sum_so_far + nums[i] <= target):
#                     so_far.append(nums[i])
#                     self.helper(so_far, i+1, nums, target, all_solns)
#                     so_far.pop()
#         return

File: backtracking/combination_sum_ii.rb
# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum2(candidates, target)
    return [] if !candidates || candidates.empty?
    @result = []
    # Sorting is to avoid the repeated combinations in the form like [1 2 3], [3 2 1]
    candidates.sort!()
    bt(candidates, 0, [], target)

    @result
end

def bt(nums, idx, path, target)
    return @result.push(path) if target == 0                         # Base case: if path sum == target, it as a solution so push & return

    idx.upto(nums.size - 1) do |i|
        break if nums[i] > target
        next if i != idx && nums[i] == nums[i - 1]                  # Skip duplicates, after i crosses idx
        bt(nums, i + 1, path + [nums[i]], target - nums[i]) # Check for solutions for target - curr number,
                                                                    # Use i + 1 as start index for next level, to avoid repition in combination
    end
end


# 40. Combination Sum II
# https://leetcode.com/problems/combination-sum-ii/description/

# Similar problem list
# https://leetcode.com/problems/combination-sum/discuss/16502/A-general-approach-to-backtracking-questions-in-Java-(Subsets-Permutations-Combination-Sum-Palindrome-Partitioning)
# https://leetcode.com/problems/combination-sum/discuss/16496/Accepted-16ms-c++-solution-use-backtracking-easy-understand.

# Points to remember
# 1. No duplicates combinations in result               (So we should consider each unique combination only once)
# 2. Can't pick same element twice Combination can contain a number only once         (This means combination without repitition)
# 3. All numbers are +ve                                (Allows for pruning totals > target)

# Approach 1: Backtracking
# Example: candidates = [2,5,2,1,2], target = 5, expected answer = [ [1,2,2], [5] ]
# Short recursion of paths only
# path: [] idx: 0
#   path: [1] idx: 1
#     path: [1, 2] idx: 2
#       path: [1, 2, 2] idx: 3
#       path: [1, 2, 5] idx: 5
#     path: [1, 5] idx: 5
#   path: [2] idx: 2
#     path: [2, 2] idx: 3
#       path: [2, 2, 2] idx: 4
#       path: [2, 2, 5] idx: 5
#     path: [2, 5] idx: 5
#   path: [5] idx: 5

# Detailed recursion tree with path, index, target and result, and pruned paths
# bt call with idx: 0 path: [] target: 5 result: []
# checking next candidate for use or prune idx: 0, candidate: 1
#     bt call with idx: 1 path: [1] target: 4 result: []
#     checking next candidate for use or prune idx: 1, candidate: 2
#         bt call with idx: 2 path: [1, 2] target: 2 result: []
#         checking next candidate for use or prune idx: 2, candidate: 2
#             bt call with idx: 3 path: [1, 2, 2] target: 0 result: []
#             bt retn with idx: 3 path: [1, 2, 2] target: 0 result: []        - Pushing to result and returning cuz target achieved
#         checking next candidate for use or prune idx: 3, candidate: 2       - Duplicate path pruned
#         checking next candidate for use or prune idx: 4, candidate: 5
#         bt retn with idx: 2 path: [1, 2] target: 2 result: [[1, 2, 2]]
#     checking next candidate for use or prune idx: 2, candidate: 2           - Duplicate path pruned
#     checking next candidate for use or prune idx: 3, candidate: 2           - Duplicate path pruned
#     checking next candidate for use or prune idx: 4, candidate: 5
#     bt retn with idx: 1 path: [1] target: 4 result: [[1, 2, 2]]
# checking next candidate for use or prune idx: 1, candidate: 2
#     bt call with idx: 2 path: [2] target: 3 result: [[1, 2, 2]]
#     checking next candidate for use or prune idx: 2, candidate: 2
#         bt call with idx: 3 path: [2, 2] target: 1 result: [[1, 2, 2]]
#         checking next candidate for use or prune idx: 3, candidate: 2
#         bt retn with idx: 3 path: [2, 2] target: 1 result: [[1, 2, 2]]
#     checking next candidate for use or prune idx: 3, candidate: 2
#     checking next candidate for use or prune idx: 4, candidate: 5
#     bt retn with idx: 2 path: [2] target: 3 result: [[1, 2, 2]]
# checking next candidate for use or prune idx: 2, candidate: 2
# checking next candidate for use or prune idx: 3, candidate: 2
# checking next candidate for use or prune idx: 4, candidate: 5
#     bt call with idx: 5 path: [5] target: 0 result: [[1, 2, 2]]
#     bt retn with idx: 5 path: [5] target: 0 result: [[1, 2, 2]]
# bt retn with idx: 0 path: [] target: 5 result: [[1, 2, 2], [5]]


require 'test/unit'
extend Test::Unit::Assertions

assert_equal(combination_sum2([10,1,2,7,6,1,5], 8), [
    [1, 7],
    [1, 2, 5],
    [2, 6],
    [1, 1, 6]
])
assert_equal(combination_sum2([2,5,2,1,2], 5), [
    [1,2,2],
    [5]
])


# **** WITH LOGGING TO SEE RECURSION TREE ***
# def combination_sum2(candidates, target)
#     return [] if !candidates || candidates.empty?
#     # Sorting is to make sure that each output list is sorted to
#     # avoid the repeated combinations in the form like [1 2 3], [3 2 1]
#     candidates.sort!()
#     bt(candidates, 0, [], [], target)
# end

# def bt(candidates, idx, path, result, target, level = -1)
#     level += 1
#     puts "#{"\t" * level}bt call with idx: #{idx} path: #{path} target: #{target} result: #{result}"
#     if target < 0
#         puts "#{"\t" * level}bt retn with idx: #{idx} path: #{path} target: #{target} result: #{result}"
#         return
#     end
#     if target == 0
#         puts "#{"\t" * level}bt retn with idx: #{idx} path: #{path} target: #{target} result: #{result}"
#         return result.push(path)
#     end

#     idx.upto(candidates.size - 1) do |i|
#         puts "#{"\t" * level}checking next candidate for use or prune idx: #{i}, candidate: #{candidates[i]}"
#         next if i != idx && candidates[i] == candidates[i - 1]                                  # Skip duplicates, after i crosses idx
#         break if nums[i] > target
#         bt(candidates, i + 1, path + [candidates[i]], result, target - candidates[i], level)    # Check for solutions for target - curr number,
#                                                                                                 # Use i + 1 as start index for next level, to avoid repition in combination
#     end
#     puts "#{"\t" * level}bt retn with idx: #{idx} path: #{path} target: #{target} result: #{result}"

#     result
# end
